File: functions/isprime64.py
def quick_p(b, p, mod):
    r = 1
    while p:
        if p % 2:
            r *= b
            r %= mod
        b *= b
        b %= mod
        p //= 2
    return r % mod


def is_prime_64(v):
    assert(v < 2**64)
    if v == 2:
        return True
    if v < 2:
        return False
    if v % 2 == 0:
        return False
    test_g = [2, 325, 9375, 28178, 450775, 9780504, 1795265022]
    startp = v-1
    while startp % 2 == 0:
        startp //= 2
    for i in range(len(test_g)):
        p = startp
        base = test_g[i]
        if base % v==0:
            continue
        find = False
        r = quick_p(base, p, v)
        while p != v-1:
            if r == v-1:
                find = True
                break
            #  -1 开始的序列, 或全1序列
            if r == 1:
                if p == startp:
                    find = True
                    break
                return False
            p *= 2
            r *= r
            r %= v
        if not find:
            return False
    return True


def main():
    print(is_prime_64(7))


main()

File: functions/isprime64.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef uint64_t ull;
#define rep(i,a,n) for (ll i=a;i<(ll)n;i++)

// Care overflow
typedef __int128_t lll;
ll quick_p(lll b, ll p,ll mod){
  lll r = 1;
  while(p){
    if(p%2)(r*=b)%=mod;
    (b*=b)%=mod;
    p/=2;
  }
  return r%mod;
}

bool miller_robin(ll v, ll base, ll startpwr){
  lll r = quick_p(base,startpwr,v);
  for(ll p = startpwr; p < v-1; p *=2){
    if(r == v-1) return true; // -1 开始的序列
    if(r == 1) return p == startpwr; // 全1序列
    (r*=r)%=v;
  }
  return false;
}

bool is_prime_64(ll v){
  if(v == 2)return true;
  if(v < 2 || v % 2 == 0)return false;
  ll p = v-1;
  while(p % 2 == 0) p /= 2;
  for(auto base:{2, 325, 9375, 28178, 450775, 9780504, 1795265022}){
    if(base % v == 0) continue; // don't break may cause 4033 bug
    // 需要所有都能找到-1开始, 或奇数次开始 全1
    if(!miller_robin(v,base,p)) return false;
  }
  return true;
}

int main(){
  vector<ll> arr = {3,5,11,15,17,19,30,325,1234,2047,4033/*37*109 */,998244353,1000000007,1000000009};
  vector<bool> res = {1,1,1,0,1,1,0,0,0,0,0,1,1,1};
  rep(i,0,arr.size()){
    auto item = arr[i];
    bool r = is_prime_64(item);
    printf("%lld:\t%d \t%s\n",item, r, r == res[i] ? "OK" :"Failed");
  }
  return 0;
}

